File: src/building_story.py
"""Turns a building's raw HPD violation records into a six-dimension profile
and an evidence-based narrative sentence.

This is the reusable version of the analysis done ad hoc in
scripts/phase3_derive_taxonomy.py, and it's the core of the "Building Story
Engine" from docs/story-taxonomy.md. Deterministic, rule-based — no ML.
"""
from collections import defaultdict
from dataclasses import dataclass, field
from datetime import datetime

NON_COMPLIANCE_STATUSES = {"NOT COMPLIED WITH", "FALSE CERTIFICATION", "INVALID CERTIFICATION"}
ACCEPTED_CERT_STATUSES = {"NOV CERTIFIED ON TIME", "NOV CERTIFIED LATE"}
P90_OVERDUE_DAYS = 9.7 * 365
P99_OVERDUE_DAYS = 25.2 * 365
MIN_CERT_ATTEMPTS_FOR_ENGAGEMENT = 3

# OrderNumbers that represent recurring ADMINISTRATIVE/FILING obligations rather
# than physical defects. These recur by design (e.g. annually, for every subject
# building) and would falsely inflate Persistent/Chronic pattern detection if not
# excluded. Found via Phase 3 testing: 8 of 25 "Persistent" matches in an 80-building
# sample turned out to be the same annual-filing code (1507). Audited the other top-60
# codes by citywide frequency (data/ordernumber_counts.json) for the same pattern —
# high confidence on the first 5, moderate on the last 2 (posting/certification-type
# obligations that plausibly recur on a compliance calendar, but not as certain as the
# bedbug report). Not exhaustive — only the top 60 of 396 codes were reviewed.
ADMINISTRATIVE_ORDERNUMBERS = {
    "780",   # "OWNER FAILED TO FILE A VALID REGISTRATION STATEMENT" - recurs if unregistered
    "1507",  # "FILE ANNUAL BEDBUG REPORT" - recurs yearly by design
    "700",   # "POST A PROPER NOTICE OF SMOKE DETECTOR REQUIREMENTS" - signage, not the device
    "1501",  # "POST A PROPER NOTICE OF CARBON MONOXIDE DETECTING DEVICE REQUIREMENTS" - signage
    "778",   # "POST AND MAINTAIN A PROPER SIGN...SHOWING THE REGISTRATION NUMBER" - signage
    "484",   # "PROVIDE A COMPLETED CERTIFICATE OF INSPECTION VISITS" - MDL S329 posting (moderate confidence)
    "623",   # "CERTIFY COMPLIANCE WITH LEAD-BASED PAINT HAZARD CONTROL REQUIREMENTS" - Local Law 1 annual cert (moderate confidence)
}


def _parse_date(s):
    if not s:
        return None
    try:
        return datetime.fromisoformat(s.replace(".000", ""))
    except Exception:
        return None


@dataclass
class BuildingProfile:
    buildingid: str
    address: str
    active_count: int
    recent_count: int
    recency_ratio: float
    class_c_total: int
    class_c_recent: int
    class_c_rate: float
    non_compliance_total: int
    non_compliance_recent: int
    accepted_cert: int
    rejected_cert: int
    cert_acceptance_rate: float | None
    max_days_overdue: int
    max_years_overdue: float
    top_sig_notices: int
    top_sig_span_years: float
    n_persistent_sigs: int
    n_chronic_sigs: int

    # Assigned dimension levels
    scale: str = ""
    recency: str = ""
    severity: str = ""
    engagement: str = ""
    pattern: str = ""
    backlog_age: str = ""


def _level_scale(n):
    if n <= 3:
        return "Minimal"
    if n <= 7:
        return "Low"
    if n <= 18:
        return "Moderate"
    if n <= 66:
        return "Large"
    return "Severe"


def _level_recency(ratio):
    if ratio < 0.15:
        return "Dormant"
    if ratio <= 0.70:
        return "Mixed"
    return "Active surge"


def _level_severity(rate):
    if rate < 0.10:
        return "Low"
    if rate <= 0.30:
        return "Elevated"
    if rate <= 0.70:
        return "Severe"
    return "Extreme"


def _level_engagement(accepted, rejected):
    attempts = accepted + rejected
    if attempts < MIN_CERT_ATTEMPTS_FOR_ENGAGEMENT:
        # Zero attempts: genuinely untested. 1-2 attempts: too little evidence
        # for a confident behavioral claim, even though technically non-zero.
        return "Untested"
    rate = accepted / attempts
    if rate < 0.30:
        return "Resistant"
    if rate <= 0.70:
        return "Mixed engagement"
    return "Responsive"


def _level_pattern(n_persistent, n_chronic):
    if n_chronic > 0:
        return "Chronic"
    if n_persistent > 0:
        return "Persistent"
    return "Isolated"


def _level_backlog(days):
    years = days / 365
    if years < 2:
        return "Current"
    if years <= 9.7:
        return "Aging"
    if years <= 25:
        return "Very aged"
    return "Extreme"


def build_profile(buildingid: str, violations: list[dict], today: datetime) -> BuildingProfile:
    """Compute the six-dimension profile for one building from its raw violation rows."""
    seen = set()
    deduped = []
    for v in violations:
        key = (v.get("apartment"), v.get("novdescription"), v.get("novissueddate"))
        if key not in seen:
            seen.add(key)
            deduped.append(v)

    addr = f"{violations[0].get('housenumber','')} {violations[0].get('streetname','')}, {violations[0].get('boro','')}"
    active_count = len(deduped)

    recent_count = class_c_recent = class_c_total = 0
    non_compliance_total = non_compliance_recent = 0
    accepted_cert = rejected_cert = 0
    max_days_overdue = 0
    signatures = defaultdict(dict)

    for v in deduped:
        nov_date = _parse_date(v.get("novissueddate"))
        cls = v.get("class")
        status = v.get("currentstatus")
        is_recent = bool(nov_date and (today - nov_date).days <= 365)

        if is_recent:
            recent_count += 1
            if cls == "C":
                class_c_recent += 1
            if status in NON_COMPLIANCE_STATUSES:
                non_compliance_recent += 1
        if cls == "C":
            class_c_total += 1
        if status in NON_COMPLIANCE_STATUSES:
            non_compliance_total += 1
        if status in ACCEPTED_CERT_STATUSES:
            accepted_cert += 1
        if status in ("FALSE CERTIFICATION", "INVALID CERTIFICATION"):
            rejected_cert += 1

        deadline = _parse_date(v.get("newcorrectbydate")) or _parse_date(v.get("originalcorrectbydate"))
        if deadline and deadline < today:
            max_days_overdue = max(max_days_overdue, (today - deadline).days)

        sig_key = (v.get("apartment"), v.get("ordernumber"))
        novid = v.get("novid")
        if nov_date and novid and v.get("ordernumber") not in ADMINISTRATIVE_ORDERNUMBERS:
            if novid not in signatures[sig_key] or nov_date < signatures[sig_key][novid]:
                signatures[sig_key][novid] = nov_date

    recurring_sigs = []
    for sig, novid_dates in signatures.items():
        dates = list(novid_dates.values())
        if len(dates) >= 2:
            span_years = (max(dates) - min(dates)).days / 365
            recurring_sigs.append((len(dates), span_years))
    recurring_sigs.sort(key=lambda x: -x[0])

    top_sig_notices, top_sig_span = recurring_sigs[0] if recurring_sigs else (0, 0.0)
    n_persistent = sum(1 for n, s in recurring_sigs if n >= 3 and s >= 2)
    n_chronic = sum(1 for n, s in recurring_sigs if n >= 10 and s >= 5)
    cert_attempts = accepted_cert + rejected_cert

    p = BuildingProfile(
        buildingid=buildingid,
        address=addr,
        active_count=active_count,
        recent_count=recent_count,
        recency_ratio=(recent_count / active_count) if active_count else 0.0,
        class_c_total=class_c_total,
        class_c_recent=class_c_recent,
        class_c_rate=(class_c_total / active_count) if active_count else 0.0,
        non_compliance_total=non_compliance_total,
        non_compliance_recent=non_compliance_recent,
        accepted_cert=accepted_cert,
        rejected_cert=rejected_cert,
        cert_acceptance_rate=(accepted_cert / cert_attempts) if cert_attempts else None,
        max_days_overdue=max_days_overdue,
        max_years_overdue=max_days_overdue / 365,
        top_sig_notices=top_sig_notices,
        top_sig_span_years=top_sig_span,
        n_persistent_sigs=n_persistent,
        n_chronic_sigs=n_chronic,
    )
    p.scale = _level_scale(p.active_count)
    p.recency = _level_recency(p.recency_ratio)
    p.severity = _level_severity(p.class_c_rate)
    p.engagement = _level_engagement(p.accepted_cert, p.rejected_cert)
    p.pattern = _level_pattern(p.n_persistent_sigs, p.n_chronic_sigs)
    p.backlog_age = _level_backlog(p.max_days_overdue)
    return p


def generate_narrative(p: BuildingProfile) -> str:
    """Assemble an evidence-based sentence from the six dimension values.
    Every clause traces to a specific field on the profile — no characterization
    of the building or owner, only what the records show."""
    parts = []

    # Scale + recency + severity opener
    if p.recency == "Active surge":
        if p.recency_ratio >= 0.98:
            recency_phrase = "all issued in roughly the past year"
        elif p.recency_ratio >= 0.90:
            recency_phrase = f"nearly all ({p.recent_count} of {p.active_count}) issued in roughly the past year"
        else:
            recency_phrase = f"the large majority ({p.recent_count} of {p.active_count}) issued in roughly the past year"
        opener = f"A wave of {p.active_count} violation{'s' if p.active_count != 1 else ''}, {recency_phrase}"
    elif p.recency == "Dormant":
        opener = f"{p.active_count} open violation{'s' if p.active_count != 1 else ''}, with little to no activity in the past year"
    else:
        opener = f"{p.active_count} open violations, {p.recent_count} of them issued in the past year"
    if p.severity in ("Severe", "Extreme"):
        opener += f", {p.class_c_total} of them serious (Class C)"
    parts.append(opener + ".")

    # Pattern
    if p.pattern == "Chronic":
        parts.append(
            f"The same defect signature has been cited {p.top_sig_notices} times over "
            f"{p.top_sig_span_years:.1f} years."
        )
    elif p.pattern == "Persistent":
        parts.append(
            f"At least one specific problem has recurred {p.top_sig_notices} times over "
            f"{p.top_sig_span_years:.1f} years."
        )

    # Engagement
    if p.engagement == "Untested":
        if p.non_compliance_total > 0:
            parts.append("No certifications have been accepted or rejected on record, though some violations show non-compliance status.")
    elif p.engagement == "Responsive":
        parts.append(f"Every certification attempt on record has been accepted ({p.accepted_cert} of {p.accepted_cert + p.rejected_cert}).")
    elif p.engagement == "Resistant":
        parts.append(f"Certification attempts have mostly been rejected ({p.rejected_cert} of {p.accepted_cert + p.rejected_cert} on record).")
    elif p.engagement == "Mixed engagement":
        parts.append(f"Certification attempts have had mixed outcomes ({p.accepted_cert} accepted, {p.rejected_cert} rejected).")

    # Backlog age
    if p.backlog_age in ("Very aged", "Extreme"):
        parts.append(f"The oldest outstanding violation is {p.max_years_overdue:.1f} years past its correction deadline.")

    return " ".join(parts)

File: src/building_story.js
/**
 * JavaScript port of src/building_story.py - must stay in exact behavioral
 * sync with the Python version. Verified against it in
 * scripts/verify_js_port.py before being trusted in the live map.
 *
 * Dates are parsed via explicit UTC construction (not `new Date(str)`)
 * because JS treats date-time strings without a timezone as LOCAL time,
 * while Python's naive datetime just does arithmetic on the literal
 * numbers. Using Date.UTC() for both "today" and record dates keeps the
 * day-difference math identical to Python regardless of the visitor's
 * browser timezone.
 */

const NON_COMPLIANCE_STATUSES = new Set(["NOT COMPLIED WITH", "FALSE CERTIFICATION", "INVALID CERTIFICATION"]);
const ACCEPTED_CERT_STATUSES = new Set(["NOV CERTIFIED ON TIME", "NOV CERTIFIED LATE"]);
const MIN_CERT_ATTEMPTS_FOR_ENGAGEMENT = 3;

const ADMINISTRATIVE_ORDERNUMBERS = new Set([
  "780", "1507", "700", "1501", "778", "484", "623",
]);

const DESCRIPTION_STOPWORDS = new Set([
  "PROPERLY", "REPAIR", "REPLACE", "REMOVE", "MAINTAIN", "PROVIDE", "CLEAN",
  "CONDITION", "ADM", "CODE", "HMC", "SECTION", "MDL", "LAW", "REQUIRED",
  "SIMILAR", "MATERIAL", "ACCORDANCE", "DESCRIBED", "NOTICE", "VIOLATION",
  "BUILDING", "APARTMENT", "LOCATED", "ENTIRE", "WHICH", "THEREFORE",
  "SUBJECT", "ABATE", "DEFECTIVE", "BROKEN", "STORY", "FRONT", "REAR",
]);

function defectKeywords(desc) {
  if (!desc) return new Set();
  const words = desc.toUpperCase().match(/[A-Z]{4,}/g) || [];
  return new Set(words.filter(w => !DESCRIPTION_STOPWORDS.has(w)));
}

function signatureIsCoherent(descriptions) {
  const distinct = [...new Set(descriptions.filter(Boolean))];
  if (distinct.length <= 1) return true;
  const counts = new Map();
  for (const d of distinct) {
    for (const kw of defectKeywords(d)) {
      counts.set(kw, (counts.get(kw) || 0) + 1);
    }
  }
  if (counts.size === 0) return false;
  return Math.max(...counts.values()) / distinct.length >= 0.6;
}

function parseDate(s) {
  if (!s) return null;
  const m = s.match(/^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})/);
  if (!m) return null;
  const [, y, mo, d, h, mi, se] = m.map(Number);
  return new Date(Date.UTC(y, mo - 1, d, h, mi, se));
}

function daysBetween(later, earlier) {
  return Math.round((later - earlier) / 86400000); // ms per day
}

function levelScale(n) {
  if (n <= 3) return "Minimal";
  if (n <= 7) return "Low";
  if (n <= 18) return "Moderate";
  if (n <= 66) return "Large";
  return "Severe";
}

function levelRecency(ratio) {
  if (ratio < 0.15) return "Dormant";
  if (ratio <= 0.70) return "Mixed";
  return "Active surge";
}

function levelSeverity(rate) {
  if (rate < 0.10) return "Low";
  if (rate <= 0.30) return "Elevated";
  if (rate <= 0.70) return "Severe";
  return "Extreme";
}

function levelEngagement(accepted, rejected) {
  const attempts = accepted + rejected;
  if (attempts < MIN_CERT_ATTEMPTS_FOR_ENGAGEMENT) return "Untested";
  const rate = accepted / attempts;
  if (rate < 0.30) return "Resistant";
  if (rate <= 0.70) return "Mixed engagement";
  return "Responsive";
}

function levelPattern(nPersistent, nChronic, realDefectCount) {
  if (nChronic > 0) return "Chronic";
  if (nPersistent > 0) return "Persistent";
  if (realDefectCount === 0) return "No real defects";
  return "Isolated";
}

function levelBacklog(days) {
  const years = days / 365;
  if (years < 2) return "Current";
  if (years <= 9.7) return "Aging";
  if (years <= 25) return "Very aged";
  return "Extreme";
}

/**
 * @param {string} buildingid
 * @param {Array<Object>} violations - raw Socrata rows for this building
 * @param {Date} today - reference date (real "now" in production, fixed in tests)
 */
function buildProfile(buildingid, violations, today) {
  const seen = new Set();
  const deduped = [];
  for (const v of violations) {
    const key = `${v.apartment || ""}|${v.novdescription || ""}|${v.novissueddate || ""}`;
    if (!seen.has(key)) {
      seen.add(key);
      deduped.push(v);
    }
  }

  const first = violations[0] || {};
  const address = `${first.housenumber || ""} ${first.streetname || ""}, ${first.boro || ""}`;
  const activeCount = deduped.length;
  const realDefectCount = deduped.filter(v => !ADMINISTRATIVE_ORDERNUMBERS.has(v.ordernumber)).length;

  let recentCount = 0, classCRecent = 0, classCTotal = 0;
  let nonComplianceTotal = 0, nonComplianceRecent = 0;
  let acceptedCert = 0, rejectedCert = 0;
  let maxDaysOverdue = 0;
  // Building-wide signatures (same OrderNumber, ANY apartment) - see the
  // matching comment in building_story.py for why apartment-scoped grouping
  // was replaced rather than kept alongside this.
  const signatures = new Map();     // ordernumber -> Map(novid -> earliest date)
  const sigApartments = new Map();  // ordernumber -> Set(apartments touched)
  const sigDescriptions = new Map(); // ordernumber -> Map(novid -> description)

  for (const v of deduped) {
    const novDate = parseDate(v.novissueddate);
    const cls = v.class;
    const status = v.currentstatus;
    const isRecent = !!(novDate && daysBetween(today, novDate) <= 365);

    if (isRecent) {
      recentCount++;
      if (cls === "C") classCRecent++;
      if (NON_COMPLIANCE_STATUSES.has(status)) nonComplianceRecent++;
    }
    if (cls === "C") classCTotal++;
    if (NON_COMPLIANCE_STATUSES.has(status)) nonComplianceTotal++;
    if (ACCEPTED_CERT_STATUSES.has(status)) acceptedCert++;
    if (status === "FALSE CERTIFICATION" || status === "INVALID CERTIFICATION") rejectedCert++;

    const deadline = parseDate(v.newcorrectbydate) || parseDate(v.originalcorrectbydate);
    if (deadline && deadline < today && !ACCEPTED_CERT_STATUSES.has(status)) {
      maxDaysOverdue = Math.max(maxDaysOverdue, daysBetween(today, deadline));
    }

    const ordernumber = v.ordernumber;
    const novid = v.novid;
    if (novDate && novid && !ADMINISTRATIVE_ORDERNUMBERS.has(ordernumber)) {
      if (!signatures.has(ordernumber)) {
        signatures.set(ordernumber, new Map());
        sigDescriptions.set(ordernumber, new Map());
      }
      const novidMap = signatures.get(ordernumber);
      if (!novidMap.has(novid) || novDate < novidMap.get(novid)) {
        novidMap.set(novid, novDate);
        sigDescriptions.get(ordernumber).set(novid, v.novdescription);
      }
      if (v.apartment) {
        if (!sigApartments.has(ordernumber)) sigApartments.set(ordernumber, new Set());
        sigApartments.get(ordernumber).add(v.apartment);
      }
    }
  }

  const recurringSigs = [];
  for (const [ordernumber, novidMap] of signatures.entries()) {
    const dates = [...novidMap.values()];
    if (dates.length >= 2) {
      const spanYears = daysBetween(
        new Date(Math.max(...dates)), new Date(Math.min(...dates))
      ) / 365;
      const breadth = (sigApartments.get(ordernumber) || new Set()).size;
      const coherent = signatureIsCoherent([...sigDescriptions.get(ordernumber).values()]);
      recurringSigs.push([dates.length, spanYears, breadth, coherent]);
    }
  }
  recurringSigs.sort((a, b) => b[0] - a[0]);

  const [topSigNotices, topSigSpan, topSigBreadth, topSigCoherent] = recurringSigs[0] || [0, 0.0, 0, true];
  const nPersistent = recurringSigs.filter(([n, s]) => n >= 3 && s >= 2).length;
  const nChronic = recurringSigs.filter(([n, s]) => n >= 10 && s >= 5).length;
  const certAttempts = acceptedCert + rejectedCert;

  const p = {
    buildingid,
    address,
    active_count: activeCount,
    real_defect_count: realDefectCount,
    recent_count: recentCount,
    recency_ratio: activeCount ? recentCount / activeCount : 0.0,
    class_c_total: classCTotal,
    class_c_recent: classCRecent,
    class_c_rate: activeCount ? classCTotal / activeCount : 0.0,
    non_compliance_total: nonComplianceTotal,
    non_compliance_recent: nonComplianceRecent,
    accepted_cert: acceptedCert,
    rejected_cert: rejectedCert,
    cert_acceptance_rate: certAttempts ? acceptedCert / certAttempts : null,
    max_days_overdue: maxDaysOverdue,
    max_years_overdue: maxDaysOverdue / 365,
    top_sig_notices: topSigNotices,
    top_sig_span_years: topSigSpan,
    top_sig_breadth: topSigBreadth,
    top_sig_coherent: topSigCoherent,
    n_persistent_sigs: nPersistent,
    n_chronic_sigs: nChronic,
  };
  p.scale = levelScale(p.active_count);
  p.recency = levelRecency(p.recency_ratio);
  p.severity = levelSeverity(p.class_c_rate);
  p.engagement = levelEngagement(p.accepted_cert, p.rejected_cert);
  p.pattern = levelPattern(p.n_persistent_sigs, p.n_chronic_sigs, p.real_defect_count);
  p.backlog_age = levelBacklog(p.max_days_overdue);
  // Independent of pattern (recurrence) - see the matching comment in
  // building_story.py for why this isn't folded into levelPattern().
  p.long_unresolved = (
    p.recency === "Dormant" &&
    p.engagement === "Untested" &&
    (p.backlog_age === "Very aged" || p.backlog_age === "Extreme")
  );
  return p;
}

function generateNarrative(p) {
  const parts = [];

  let opener;
  if (p.recency === "Active surge") {
    let recencyPhrase;
    if (p.recency_ratio >= 0.98) {
      recencyPhrase = "all issued in roughly the past year";
    } else if (p.recency_ratio >= 0.90) {
      recencyPhrase = `nearly all (${p.recent_count} of ${p.active_count}) issued in roughly the past year`;
    } else {
      recencyPhrase = `the large majority (${p.recent_count} of ${p.active_count}) issued in roughly the past year`;
    }
    opener = `A wave of ${p.active_count} violation${p.active_count !== 1 ? "s" : ""}, ${recencyPhrase}`;
  } else if (p.recency === "Dormant") {
    opener = `${p.active_count} open violation${p.active_count !== 1 ? "s" : ""}, with little to no activity in the past year`;
  } else {
    opener = `${p.active_count} open violations, ${p.recent_count} of them issued in the past year`;
  }
  if (p.severity === "Severe" || p.severity === "Extreme") {
    opener += `, ${p.class_c_total} of them serious (Class C)`;
  }
  parts.push(opener + ".");

  if (p.pattern === "No real defects") {
    parts.push(
      "None of these are physical defect records — every one is an administrative " +
      "filing requirement (such as registration or bedbug-report compliance), not a " +
      "cited problem with the building itself."
    );
  }

  if (p.pattern === "Chronic" || p.pattern === "Persistent") {
    let where;
    if (p.top_sig_breadth >= 2) {
      where = `across ${p.top_sig_breadth} different apartments`;
    } else if (p.top_sig_breadth === 1) {
      where = "in the same apartment";
    } else {
      where = "in the building's common areas";
    }
    const what = p.top_sig_coherent
      ? "The same specific problem"
      : "The same administrative code — covering several different underlying defects —";
    parts.push(`${what} has recurred ${p.top_sig_notices} times over ${p.top_sig_span_years.toFixed(1)} years, ${where}.`);
  }

  if (p.engagement === "Untested") {
    if (p.non_compliance_total > 0) {
      parts.push("No certifications have been accepted or rejected on record, though some violations show non-compliance status.");
    }
  } else if (p.engagement === "Responsive") {
    parts.push(`Every certification attempt on record has been accepted (${p.accepted_cert} of ${p.accepted_cert + p.rejected_cert}).`);
  } else if (p.engagement === "Resistant") {
    parts.push(`Certification attempts have mostly been rejected (${p.rejected_cert} of ${p.accepted_cert + p.rejected_cert} on record).`);
  } else if (p.engagement === "Mixed engagement") {
    parts.push(`Certification attempts have had mixed outcomes (${p.accepted_cert} accepted, ${p.rejected_cert} rejected).`);
  }

  if (p.backlog_age === "Very aged" || p.backlog_age === "Extreme") {
    if (p.long_unresolved) {
      parts.push(
        `The oldest outstanding violation has gone unaddressed for ${p.max_years_overdue.toFixed(1)} years ` +
        "— no certification has ever been attempted, and there's been no recent activity on record."
      );
    } else {
      parts.push(`The oldest outstanding violation is ${p.max_years_overdue.toFixed(1)} years past its correction deadline.`);
    }
  }

  return parts.join(" ");
}
